dash/include/dash/halo/HaloMatrixWrapper.h

```diff
@@ -200,104 +217,140 @@ class HaloMatrixWrapper {
+                                    dart_get_handle(off, it.dart_gptr(),
+                                                    region_size, index_type,
+                                                    ds_num_elems_block.dtype,
+                                                    &data.handle);
+                                  },
+                                  std::move(halo_data) }));
         }
 
         num_elems_block = region.region().extent(0);
       }
     }
   }
 
+  /**
+   * Constructor that takes \ref Matrix and a user
+   * defined number of stencil specifications (\ref StencilSpec).
+   * The \ref GlobalBoundarySpec is set to default.
+   */
+
+  template <typename... StencilSpecT>
+  HaloMatrixWrapper(MatrixT& matrix, const StencilSpecT&... stencil_spec)
+  : HaloMatrixWrapper(matrix, GlobBoundSpec_t(), stencil_spec...) {}
+
+  HaloMatrixWrapper() = delete;
+
   ~HaloMatrixWrapper() {
     for(auto& dart_type : _dart_types) {
       dart_type_destroy(&dart_type);
     }
     _dart_types.clear();
   }
 
-  iterator begin() noexcept { return _begin; }
-
-  const_iterator begin() const noexcept { return _begin; }
-
-  iterator end() noexcept { return _end; }
-
-  const_iterator end() const noexcept { return _end; }
-
-  iterator_inner ibegin() noexcept { return _ibegin; }
-
-  const_iterator_inner ibegin() const noexcept { return _ibegin; }
-
-  iterator_inner iend() noexcept { return _iend; }
-
-  const_iterator_inner iend() const noexcept { return _iend; }
-
-  iterator_bnd bbegin() noexcept { return _bbegin; }
-
-  const_iterator_bnd bbegin() const noexcept { return _bbegin; }
-
-  iterator_bnd bend() noexcept { return _bend; }
-
-  const_iterator_bnd bend() const noexcept { return _bend; }
-
+  /// returns the underlying \ref HaloBlock
   const HaloBlock_t& halo_block() { return _haloblock; }
 
+  /**
+   * Initiates a blocking halo region update for all halo elements.
+   */
   void update() {
     for(auto& region : _region_data)
       update_halo_intern(region.second, false);
   }
 
+  /**
+   * Initiates a blocking halo region update for all halo elements within the
+   * the given region.
+   */
   void update_at(region_index_t index) {
     auto it_find = _region_data.find(index);
     if(it_find != _region_data.end())
       update_halo_intern(it_find->second, false);
   }
 
+  /**
+   * Initiates an asychronous halo region update for all halo elements.
+   */
   void update_async() {
     for(auto& region : _region_data)
       update_halo_intern(region.second, true);
   }
 
+  /**
+   * Initiates an asychronous halo region update for all halo elements within
+   * the given region.
+   */
   void update_async_at(region_index_t index) {
     auto it_find = _region_data.find(index);
     if(it_find != _region_data.end())
       update_halo_intern(it_find->second, true);
   }
 
+  /**
+   * Waits until all halo updates are finished. Only useful for asyncronous
+   * halo updates.
+   */
   void wait() {
```

Comment:
Eventually, region-specific `wait` and *test* would be nice :)

Reply:
done